### DISCUSSION TITLE
google workspace auth failure - invalid credentials

### DISCUSSION THREAD

**Discussion by mike-freislich** [imapsync/imapsync#430] [Q&A]
Hi,

I've been using imapsync to successfully migrate a user account from one google workspace to another.
I am now trying a second user in the same source domain.

as with the previous successful account migration, on the source account I have:
- setup the gmail account imap settings correctly
- enabled two factor authentication
- generated an app password

Here is imapsync 2.271 on host MYHOST, a linux system with 3.0/7.6 free GiB of RAM, 27.65% used by processes.
with Perl 5.34.0 and Mail::IMAPClient 3.43
...
Host1: IMAP server [imap.gmail.com] port [993] user [user@gworkspace_sourceDomain]
Host1: connecting and login on host1 [imap.gmail.com] port [993] with user [user@gworkspace_sourceDomain]
Host1 IP address: xxx.xxx.xxx.xxx Local IP address: yyy.yyy.yyy.yyy
Host1 banner: * OK Gimap ready for requests from xxx.xxx.xxx.xxx cg17mb112987281wrb
Host1 capability before authentication: IMAP4rev1 UNSELECT IDLE NAMESPACE QUOTA ID XLIST CHILDREN X-GM-EXT-1 XYZZY SASL-IR AUTH=XOAUTH2 AUTH=PLAIN AUTH=PLAIN-CLIENTTOKEN AUTH=OAUTHBEARER AUTH
**Host1 info: authmech [LOGIN] user [user@gworkspace_sourceDomain] authuser [] IsUnconnected []**
**Host1 failure: Error login on [imap.gmail.com] with user [user@gworkspace_sourceDomain] auth [LOGIN]: 2 NO [AUTHENTICATIONFAILED] Invalid credentials (Failure)
Host1: failed login on [imap.gmail.com] with user [user@gworkspace_sourceDomain] auth [LOGIN]**

I am stumped ... the credentials are correct... 

Any clue as to what I'm doing wrong?

**Comment by gilleslamiral** [comment]
No clue but the classical 
https://imapsync.lamiral.info/FAQ.d/FAQ.Authentication_failure.txt

**Comment by mike-freislich** [comment]
Thank you for your quick response Gilles.

It sent me on the path towards success!

OMG…. Here’s the solution!

I have migrated 3 accounts from the same source Google Workspace to 2
different Google Workspace Instances. I am using app-passwords for
authentication. Google generates the app-passwords in the following format
: “aaaa bbbb cccc dddd”. For the first two accounts this all worked
perfectly. I also needed to delete any emails larger than 30MB in order for
the IMAP-append operation to work correctly on the target workspace
accounts (google has a hard set maximum for map-append).

For the last account, authentication would simply not work. I used "openssl
s_client” to verify authentication, as per your notes. Again the password
format authenticated perfectly for the first two accounts, but not for the
third. It failed with "invalid credentials”. I took a guess, and removed
the spaces from the app-password e.g. “aaaa bbbb cccc dddd” ->
“aaaabbbbccccdddd”… viola! Solved.

Would have been nice for google to have some kind of standard here, along
with trouble shooting explanation when dishing out these app-passwords.
Anyway… all is working now.

Perhaps this will help someone else in the future.

Thanks again for your help!

-Mike

**Comment by gilleslamiral** [comment]
Thanks, Mike!

I updated the document
https://imapsync.lamiral.info/FAQ.d/FAQ.Gmail.txt
````
1) "App password with 2-step verification" is the first way to 
   authenticate to Gmail with Imapsync.

Go to
https://myaccount.google.com/security
and sign in with the account you want to transfer email to or from.
Go to the "Signing in to Google" part of the security page and follow
the process to activate the "2-step verification" until it is ok.

Then, just below "2-step verification" click on "App passwords"
and follow the instruction to get a 16 characters password, on the
pull-down button "Select app", choose "Other (Custom name)" and call
it imapsync for example.  Now use this password to access the
mailbox with imapsync. If you copy/paste the password, remove the 
blanks inside the password, the blanks are not part of the password,
they're just there to facilitate the reading. If you leave the blanks,
you may encounter an authentication error.
````

**Comment by rodrigo-norys** [comment]
Thank you guys, for the information. Helped me a lot!!!
